### ISSUE TITLE
cannot locate symbol "posix_spawn"

### ISSUE THREAD

**Issue by mipastgt** [gluonhq/substrate#741]
I cannot run my app on my Android 8 phone anymore although the same apk file still works on my Android 10 tablet. The error I get is: 

```
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] V/GraalActivity(19905): surfaceCreated for com.gluonhq.helloandroid.MainActivity@2f21180
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] V/GraalActivity(19905): loading substrate library
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] --------- beginning of crash
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): FATAL EXCEPTION: main
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): Process: de.mpmediasoft.de.mpmediasoft.mpcopilot.main.mobile, PID: 19905
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "posix_spawn" referenced by "/data/app/de.mpmediasoft.de.mpmediasoft.mpcopilot.main.mobile-kcwTzX1RCuWIT-mr_4xf-A==/lib/arm64/libsubstrate.so"...
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at java.lang.Runtime.loadLibrary0(Runtime.java:1016)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at java.lang.System.loadLibrary(System.java:1657)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at com.gluonhq.helloandroid.MainActivity.surfaceCreated(MainActivity.java:104)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.SurfaceView.updateSurface(SurfaceView.java:753)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.SurfaceView$2.onPreDraw(SurfaceView.java:154)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:1045)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2808)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1780)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7827)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.Choreographer.doCallbacks(Choreographer.java:723)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.Choreographer.doFrame(Choreographer.java:658)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.os.Handler.handleCallback(Handler.java:789)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.os.Handler.dispatchMessage(Handler.java:98)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.os.Looper.loop(Looper.java:164)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at android.app.ActivityThread.main(ActivityThread.java:6944)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at java.lang.reflect.Method.invoke(Native Method)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] E/AndroidRuntime(19905): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] W/ActivityManager( 3652): crash : de.mpmediasoft.de.mpmediasoft.mpcopilot.main.mobile,0
[Thu Aug 13 12:58:43 CEST 2020][INFO] [SUB] W/ActivityManager( 3652):   Force finishing activity de.mpmediasoft.de.mpmediasoft.mpcopilot.main.mobile/com.gluonhq.helloandroid.MainActivity
[Thu Aug 13 12:58:44 CEST 2020][INFO] [SUB] W/ActivityManager( 3652): setHasOverlayUi called on unknown pid: 19905

```

I am using `<client-maven-plugin.version>0.1.31-SNAPSHOT</client-maven-plugin.version>
`

**Comment by jperedadnr** [comment]
Can you run and post the result of:

```
nm target/client/aarch64-android/gvm/android_project/app/src/main/jniLibs/arm64-v8a/libsubstrate.so | grep posix_spawn
```
You build the project just once and then deploy it to both devices, right?

**Comment by mipastgt** [comment]
I think I have created this problem myself by my attempts to get SQLite working. See:
https://stackoverflow.com/questions/63375536/sqlite-on-android-via-javafx-and-gluons-client-maven-plugin

The SQLite library tries to load a platform specific shared library and that code seems to cause the above problem on my phone. I have now removed this code, because it does not work anyway, and now I can run my app again on phone and tablet. The command you asked me to run returns nothing now after my code removal. What makes me wonder is why this did not cause a problem on the tablet.

So I think this issue can be closed but maybe you could still answer my SO question whether loading of user provided native libraries is supported or not :-)

**Comment by jperedadnr** [comment]
Thanks, it is still a valid issue, and we have to fix it either way.

**Comment by jperedadnr** [comment]
About your SO question:
- we do support static native libraries (*.a) (i.e from Attach)
- loading of more generic native files is "in progress" (see https://github.com/gluonhq/substrate/issues/715). 

There are a few problems. For instance, the sqlite jar contains the same library *.so for many different architectures, so there is the problem of Substrate extracting the right one or all of them to different folders.

Then there is also the need to extract to the right place. The .so libraries on Android must go to `app/src/main/jniLibs/arm64-v8a` (where we already place `libsubstrate.so`).

I've tried "manually" some of the sqlite libs, none were valid (wrong architecture). They don't build for Android with 64 bits. So maybe you can try it yourself?

**Comment by mipastgt** [comment]
> There are a few problems. For instance, the sqlite jar contains the same library *.so for many different architectures, so there is the problem of Substrate extracting the right one or all of them to different folders.

I have got that working already with client-maven-plugin on all desktop platforms. I just add this line

`<list>org/sqlite/native/${sqlite.platform}/${sqlite.arch}/.*</list>`

to the `resourcesList` and set the corresponding variables. In addition I have to add this `jnilist`

```
<jniList>
    <list>org.sqlite.core.DB</list>
    <list>org.sqlite.core.NativeDB</list>
    <list>org.sqlite.BusyHandler</list>
    <list>org.sqlite.Function</list>
    <list>org.sqlite.ProgressHandler</list>
    <list>org.sqlite.Function$Aggregate</list>
    <list>org.sqlite.Function$Window</list>
    <list>org.sqlite.core.DB$ProgressObserver</list>
</jniList>
```

> Then there is also the need to extract to the right place. The .so libraries on Android must go to app/src/main/jniLibs/arm64-v8a (where we already place libsubstrate.so).

That's still an open problem.

> I've tried "manually" some of the sqlite libs, none were valid (wrong architecture). They don't build for Android with 64 bits. So maybe you can try it yourself?

Ok, so it does not make sense to further try any of these.

**Comment by jperedadnr** [comment]
Indeed, the same happens for iOS (no native library). A good exercise could be building SQLite for both Android and iOS  :)

**Comment by jperedadnr** [comment]
@mipastgt We have a possible fix for the posix_spawn issue, but we'll need to test over your failing case. Is it possible that you can get it running again?

If that is the case, please replace `~/.gluon/substrate/javaStaticSdk/11-ea+0/ios-arm64/labs-staticjdk/lib/static/libjava.a` with this file (unzip first):

[libjava.a.zip](https://github.com/gluonhq/substrate/files/5069395/libjava.a.zip)

Thanks.

**Comment by mipastgt** [comment]
I'll give it a try.

**Comment by mipastgt** [comment]
I have successfully restored the crash case but I then realized that you want me to replace the .../ios-arm64/... lib but I can only build for Android.

**Comment by jperedadnr** [comment]
My bad, the path is for android:

`~/.gluon/substrate/javaStaticSdk/11-ea+0/android-aarch64/labs-staticjdk/lib/static/libjava.a`

**Comment by mipastgt** [comment]
I was just wondering because the libjava.a was zipped on a Mac and not a Linux box. (It contained a __MACOSX folder).

Anyway, I replaced the libjava.a as you said and I can confirm that after a rebuild of my app I can now run the same apk file again on the phone and the tablet without the error which I reported above.

**Comment by jperedadnr** [comment]
(that comes from my Mac... I couldn't link the .a directly here, so I had to zip it)

Great, thanks for testing.

**Comment by jperedadnr** [comment]
Fixed with #750
